### PR TITLE
Document the initial public API design for Nullable Reference Types

### PR DESCRIPTION
**Pull request by 333fred**
Adds a document giving an overview of the public API we're targeting for exposing nullable reference types information from the compiler layer to IDE and analyzer services. This includes the initial API design but is missing some parts, especially around syntax generation. These will come later.

@dotnet/roslyn-compiler @dotnet/roslyn-ide @CyrusNajmabadi @terrajobst @mattwar please take a look.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
I'm curious the basis for:

>Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis.
This includes features like intellisense.

We have ample information, for example, that the compiler is more than fast enough, even on large methods where it has to bind everything (i.e. because you're contained in large lambdas).  Can we actually get some measurements on roslyn for the expected costs here and if that is actually a problem given the existing budget for intellisense?    I would be very wary of making design decisions based on speculative perf concerns without the data to back it up.  Thanks!

**Comment by jasonmalinowski** [comment]
> We have ample information, for example, that the compiler is more than fast enough, even on large methods where it has to bind everything (i.e. because you're contained in large lambdas). Can we actually get some measurements on roslyn for the expected costs here and if that is actually a problem given the existing budget for intellisense? I would be very wary of making design decisions based on speculative perf concerns without the data to back it up. Thanks!

We did do test where the @333fred made a change to have method binding always bind the whole method. We didn't see any obvious problems running RPS with this; we concluded that if this was actually a problem, we don't have the systems today to prove it either way.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+Nullability information will be retrieved from the `SemanticModel`, much like type information.
+`GetSemanticModel` will return a `SemanticModel` that is aware of nullability information, and will force binding and calculation of nullability information when queried.
+We will offer a method off of `SemanticModel`, `GetSingleStatementSemanticModel()`, which will return a `SemanticModel` that does not calculate nullability information, for particularly performance-sensitive scenarios.
+This `SemanticModel` will share a cache with its parent, much like speculative `SemanticModel`s, and all nullabilities accessed on symbols retrieved from this `SemanticModel` will be `NotAnalyzed`.
+
+Generally speaking, when you want to determine the nullability of an `ITypeSymbol`, you need to check the containing context.
+The declaring symbol (e.g. `ILocalSymbol`, `IFieldSymbol`, etc.) will have the declared nullability of the type.
+`GetTypeInfo` will give you both the `ITypeSymbol` and the `Nullability` of an expression.
+Nested nullabilities for type parameters are contained on the containing `ITypeSymbol`.
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: this mix seems problematic.  it means there's now two sets of data that has to be used as a 'source of truth' for nullability. We use symbols in some cases, but not all.  

Given that nullability information is stored on a symbol, how do you reconcile this with:

> These include reference equality for `ITypeSymbol`s that aren't generic, such as `string`.

While i get that this is about non-generics, how do you do equality properly *for* generics.  for example, say i have:

`IList<string?>` and `IList<string>`.  Even if do `.Equals`, will that return `true`?  If not, that seems like you will be breaking existing code that expects those to be the same.  

However, if you're ok with those symbols *not* being hte same, it's unclear to me why it's then a problem for something like `string?` and `string` to not be reference equals.

**Comment by CyrusNajmabadi** [comment]
> This includes the initial API design but is missing some parts, especially around syntax generation. These will come later.

It is hard to gauge any proposed API if we're left without seeing how this bear out in practice in such a critical scenario.  Can we have the discussion about both at the same time?  Or at least have a pair of proposals?  In the absence of seeing how we could generate code, i really dont' know how to effectively judge this given that that's a huge part of what the IDE uses these APIs for.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+If the user requests information while asking for nullability to not be calculated, we'll use this to stand in for nullability.
+* `Unknown` - This is used for legacy C# code that does not know about nullability.
+This can appear in modern C# code that calls unannotated C#, or in certain generic variance scenarios where a common nullability cannot be inferred from the supplied arguments.
+For use in this document unknown is represented with the `~` character, but this character does not appear in source.
+* `NotNull` - This is used when we know the nullability is not null.
+* `MayBeNull` - This is used when we know the nullability may be null.
+
+Nullability information will be retrieved from the `SemanticModel`, much like type information.
+`GetSemanticModel` will return a `SemanticModel` that is aware of nullability information, and will force binding and calculation of nullability information when queried.
+We will offer a method off of `SemanticModel`, `GetSingleStatementSemanticModel()`, which will return a `SemanticModel` that does not calculate nullability information, for particularly performance-sensitive scenarios.
```
Comment (terrajobst) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: > `GetSingleStatementSemanticModel()`

Is that the name we concluded we want?

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+Value types are always considered `NotNull`.
+Nullable value types and reference types have their state tracked, and are either `NotNull` or `MayBeNull` depending on flow state.
+Unconstrained generic parameters are being tracked, but what they will return for flow state information before explicit null checks occur in source code has not yet been decided.
+
+### Declared Nullability
+
+Declared nullability is the nullability that was explicitly typed in the source code by the programmer.
+There are some cases where this is inferred, such as lambda arguments/return types, `var` variables, and type parameter substitution.
+For these cases, flow state is necessary to compute the declared nullability.
+For example:
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: i find this a bit hard to parse.  "explicitly typed in code" "some cases where this is inferrred".  If it's inferred, it doesn't seem like it was explicitly typed in code.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+`CandidateSymbols` will not have calculated nullabilities, as we cannot run nullable analysis on code that has failed overload resolution.
+
+### `GetSemanticModel(bool skipNullabilityInformation)`
+
+This API proposal changes the way `SemanticModel` calls behave by default.
+Currently, if semantic information is requested for a `SyntaxNode`, we make an effort to bind as little as possible to provide semantic information about that statement, usually a single statement.
+However, if that semantic information will include `Nullability`, we must bind all statements that come before that statement and run nullability analysis up to and including that statement.
+We believe that this change will be ok for most scenarios, as things such as colorization and analyzers are already running in the background near-constantly, binding pretty much everything on screen and in the file anyway.
+There are a few specific scenarios that will be perf-sensitive, and an overload for `GetSemanticModel()` that skips nullability information will be provided for these.
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: this seems speculative.  why not see how far the above gets us, and add this as necessary.  in the IDE's experience, perf is already so good (with ambient features causing near-entire-file binding) that it's unclear to me if we'd need this.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    MayBeNull
+}
+
+#region Declared Nullability
+
+// These can potentially be influenced by calculated flow state, but do not change after initial declaration
+
+public interface IDiscardSymbol : ISymbol
+{
+    Nullability Nullability { get; }
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: this seems weird, given that you can't really interact with a discarded symbol right?  So when does this Nullability ever come into play?

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 2 above)
Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: Whoops, no, must have missed a spot.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+public interface ILocalSymbol : ISymbol
+{
+    Nullability Nullability { get; }
+}
+
+public interface IMethodSymbol : ISymbol
+{
+    Nullability ReturnNullability { get; }
+    ImmutableArray<Nullability> TypeArgumentsNullabilities { get; }
+    Nullability ReceiverNullability { get; }
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: this feels super unpleasant.  We appear to be making a *dual* API where you need to have an entirely parallel set of nullability info that goes along with any location in our symbolic API where we might reference a type.  This seems onerous and unpleasant, and it is unlike how we did anything else in teh API.

For example, we do not have:
"ReturnIsPointer", "ReturnIsArray", "ReturnIsTuple", "ReturnIsDynamic" or 
`ImmutableArray<bool> TypeArgumentsAreArrays`, or `ImmutableArray<bool> TypeArgumentsAreDynamic` or `ImmutableArray<TupleNameInfo> TypeArgumentTupleNames`, etc. etc.

Instead of using parallel construction, we have always just conveyed different types as actual different types in roslyn.  This violates that construction.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+#region Printing
+
+public interface ITypeSymbol
+{
+    string ToDisplayString(Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToDisplayParts(Nullability nullability, SymbolDisplayFormat format = null);
+    string ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+}
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: this feels like the tip of the iceberg.  Because ITypeSymbol doesn't just know if it's nullable, this information must be passed along.  Note: this is viral.  Any method that needs to bottom out here will need to pass tha tinformation along.  Similarly, many existing methods will now need to take ITypeSymbols *and* this new 'Nullability' info.  This too will be viral, necessitating large swaths of code be updated accordingly.  For existing APIs that are ITypeSymbol-based, this seems like this will effectively be deprecating things.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+/*
+ * There will be more API changes, such as on SyntaxFactory and SyntaxGenerator. This section will be updated later when we have designs for this ready.
+ */
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: This seems necessary to do now.  Especially so we can gauge how bad the viral nature of this new piece of data would be.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+## Considered Alternatives
+
+### `Nullability` on `ITypeSymbol`
+
+The basic principle of this alternative proposal is to add `Nullability Nullability { get; }` to `ITypeSymbol`, and remove the rest of the proposed nullability retrieval APIs in the main section.
+This has a few benefits:
+
+1. Consumers of `ITypeSymbol`s won't have to keep track of where the type symbol came from to retrieve nullability information, they just have to ask the type.
+2. Printing APIs and code generation APIs do not need to have new overloads introduced to allow communication of top-level nullability information.
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: note: i think '2' is being understated.  It's not just these APIs.  It's *any* api that transitively ends up using tehse APIs.  This is fairly significant.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+3. Consumers of the new API don't have to separately worry about nullability when comparing types, as `string`, `string~`, and `string?` are truly different types.
+
+However, this approach has a few major issues that ultimately break our ability to do this design.
+
+1. This approach will violate principle #1 of our goals for this API, breaking existing analyzers and refactorings extensively.
+A common strategy for these APIs is to call `Compilation.GetTypeByMetadataName`, or one of the similar overloads, and then look for fields, locals, properties, or other similar symbols of the requested type.
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: Can you explain this.  Why would an existing analyzer/refactoring break?  

> A common strategy for these APIs is to call `Compilation.GetTypeByMetadataName`, or one of the similar overloads, and then look for fields, locals, properties, or other similar symbols of the requested type.

These would not break.  Those fields are not of that type.   If i had a field of type 'string' before then it will still be found.  But if i change my API to use `string?` then the analyzer should *not* find it since htat's a different type, and there is no reason to believe that analyzer/refactoring will work on that type properly.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 4 above)
Comment (terrajobst) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: From an API standpoint, I think the evolution of `GetSemanticModel()` is the right balance (i.e. the existing API everybody uses today will just bind more stuff, but we believe perf is acceptable and that providing higher-fidelity to consumers is more important). From an API-design standpoint overloads are almost "free" in that that they have been proven to not increase the concept count for consumers, assuming the ones with fewer arguments work in the majority of cases, which seem to be the case here.

I'm not qualified to argue the perf issues with the new model and whether or not that adding the the new API is necessary or not. However, given that you can update Roslyn fairly easily, I would urge you to consider @CyrusNajmabadi proposal to exclude the API unless you have compelling evidence that the API is needed. The risk in giving an API prematurely is that by the time you understand the impacted scenario you realized that the API you have provided isn't the right one either and you need to expose a different switch (i.e. not `skipNullabilityInformation` but `skipFizBuzzAnalysis`). If the opportunity cost of doing the API later is tiny, deferring is the better approach IMHO.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+Up to this point, our guidance has been that, for non-generic types, it is perfectly fine to just compare these symbols via `==` if you're looking for exact matches, and that you only need to resort to `OriginalDefinition` for types with generics.
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: i don't believe that's ever been the guidance.  For example, evne in Roslyn itself we continuously tell people that they need ot use .Equals even on non-generics since there is no guarantee of any sort of reference equality.

That said, even if users use .Equals i would *still* expect these types not to be the same.  `int?` is not the same type as `int` regardless of whether i use `==` or `.Equals`.  Similarly, `dynamic` is not the same as `object` and `(int, int)` is not the same as `(int a, int b)`.  These are different types, and my expectation as an analyzer author is that i should not consider the same.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+This will no longer be possible if we were to add `Nullability` to `ITypeSymbol`.
+2. The default `Nullability` of symbols is another pain point.
+When a class is declared, it doesn't have a nullability.
+It's just an `INamedTypeSymbol`.
+It may have constraints that have nullabilities, but the top-level nullability is not a question that makes any sense.
+However, in order to not break analyzers on nullable-unaware code, we will have to make the default nullability `Unknown`, as old C# code will have this as the nullability when `GetTypeInfo` is called on an expression.
+This would further exacerbate issue #1.
+3. This causes massive explosion in type heirarchies.
+If a user got `string?` and called `GetDeclaredMembers()`, the parent of each of these members must point back to `string?`, and not `string` or `string~`.
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: this sounds *correct*, and what i would expect.  For example, when i call .GetMembers on a `(int i, int j)` i get a different set of members than what i get on `(int x, int y)`.  Similarly, if i get members on any instantiated generic, i get a different set of members.  Same with thigns like arrays, pointers, etc.  They are different types, so i get different results.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+
+For these reasons, we have decided not take this approach.
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: These reasons don't seem to make a lot of sense to me.  They seem very speculative and are optimizing for a strange view of the world that seems to want to convey that these types are the same, even though the language is heavily pushing the intuition that they're not.  I would have expected this approach if we had gone with, say, an `[Nullable]`-attribute based way of representing this.  But not when we go the path of something like `string?`.  

Note: it's exceptionally strange since we already shipped the `type?` feature already with nullable-value-types.  And We certainly didn't use the `Nullable` enum for that.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+An early alternative proposal had a parallel API to `GetTypeInfo()`, `GetNullabilityInfo()`, that would return a structure similar to `TypeInfo` that contained nullability information.
+We decided that this would provide no benefits: because of nested nullability, we would still need to run nullability analysis on calls to `GetTypeInfo`, and it would needlessly separate information that users will want to get access to in tandem.
+Additionally, we could not come up with a good set of APIs for `GetDeclaredSymbol` and `GetSymbolInfo` to go along with that API.
+
+### Lazily calcualted `Nullability` properties
+
+This alternative proposal added similar `Nullability` properties to the interfaces as the main proposal, but did not calculate them ahead of time.
+Rather, they would be calculated when requested by the user.
+The big problem with this is that suddenly symbols, `TypeInfo`, and `SymbolInfo` must suddenly start carrying around the context in which they were called to be able to lazily calculate these properties, which is a major breaking change.
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: can you clarify how that would be a breaking change?  Wouldn't it be an internal detail that users wouldn't know about?

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+
+### `ITypeReference`
+
+We briefly wandered down a thought experiment where we considered adding an `ITypeReference` API, which would contain an `ITypeSymbol` and a `Nullability`, plus potentially things like ref, custom modifiers, and other side car information we have today.
+This would be an *extremely* breaking change, requiring massive changes across the API surface area.
+It would also require revisiting an explicit design choice in the initial design of Roslyn.
```
Comment (CyrusNajmabadi) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: i agree. this doesn't seem like a good idea.

**Comment by CyrusNajmabadi** [comment]
To follow this thought experiment:

> It seems like we're worried about them... but it doesn't seem like a problem that's particularly worrying.

Can we walk through teh analyzer/refactoring concern?  Say we introduce Nullability on ITypeSymbols.  What does hte exercise look like to update an existing analyzer/refactoring?  Can we try it on our own stuff to see what it's like?  Is it actually difficult or costly to do?  I would actually expect that it would not be htat hard to consume, and testing our stuff out on just 3-4 nullable cases would indicate the problems in the code, and would only take minor tweaks to fix.

This seems far more palatable to me than having to go deprecate large swaths of our API surface area, build a new set that then matches this new symbolic data hierarchy, and then have people start using that in order for things to work properly.

**Comment by CyrusNajmabadi** [comment]
As another point of comparison.  TypeScript supports very complex flow-inferred types.  For example, For a string, it can understand and represent different combinations of the following states:

1. undefined
2. null
3. empty
4. not-empty

So `string` is actually a way of saying `empty | not-empty`.  Similarly, `string?` means `string | undefined`.  But you can also have things like `string | null`, `string | null | undefined`.  And code like `if (s)` will end up saying the string is `non-empty` in teh true branch, but `undefined | null | empty` in the false branch. 

From my understanding, this sort of precise type-system information (where these differences are stored as unique types) has not been a problem for them either in terms of performance, or in terms of issues with their public analyzer model.  TypeScript also has a thriving analyzer ecosystem, and they've added all sort of interesting type-related stuff to it over its lifetime (including nullability info) without needing to contort the API for these concerns.  Given the existence proof of them doing basically exactly what we're doing here, i'm very skeptical we need to take this approach.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+We decided against this because we believe that, except for specific cases where perf is an issue, the savings gained from not calculating nullability would be moot as IDE features or analyzers will have already requested nullability anyway, and the time savings would be unused.
+Therefore, we are not taking this approach.
+
+### `GetNullabilityInfo`
+
+An early alternative proposal had a parallel API to `GetTypeInfo()`, `GetNullabilityInfo()`, that would return a structure similar to `TypeInfo` that contained nullability information.
+We decided that this would provide no benefits: because of nested nullability, we would still need to run nullability analysis on calls to `GetTypeInfo`, and it would needlessly separate information that users will want to get access to in tandem.
+Additionally, we could not come up with a good set of APIs for `GetDeclaredSymbol` and `GetSymbolInfo` to go along with that API.
+
+### Lazily calcualted `Nullability` properties
```
Comment (jasonmalinowski) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]:
```suggestion
### Lazily calculated `Nullability` properties
```

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 10 above)
Comment (CyrusNajmabadi) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: But, to do that, it would need to do .OriginalDefinition anyways.  Because `Task<T>` won't match `Task<string>` or `Task<int>` etc. etc.  So it already needs to deal with variadic concept of type matching.

**Comment by 333fred** [comment]
@CyrusNajmabadi thanks for the continued and effusive feedback here. I know it seems like we're being stubborn here, but I want to make clear that we seriously considered putting `Nullability` directly on `ITypeSymbol`, and over several hours of design time have decided that that approach is not how we want to go with API.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 3 above)
Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: I'll add a `usually` qualifier. Declared nullability _is_ usually explicitly typed in code, but we can infer it wherever we infer a type.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 4 above)
Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: I'll add a note about this that we'll only provide this API if dogfooding shows that it's necessary.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 5 above)
Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: You can't interact with it, but the IDE will need to display something for QuickInfo when the user hovers over it so we need to provide the info.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 6 above)
Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: This is very similar to how we do both `ref` and custom modifiers today, that's the model we're taking here.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 14 above)
Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: TypeInfo is currently a light wrapper type around two ITypeSymbols. The IDE hangs onto these all the time, and they don't have any references back to the context (read, position in a source tree) they came from. In order to calculate nullability information, you must have a location in a source tree, so these structs would suddenly no longer be lightweight wrappers.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi thanks for the continued and effusive feedback here. I know it seems like we're being stubborn here, but I want to make clear that we seriously considered putting Nullability directly on ITypeSymbol, and over several hours of design time have decided that that approach is not how we want to go with API.

Understood.  Though something seems a bit baffling to me.  Having read through the proposal, and notes, and conversations between the two of us, it seems like a lot of this is simply coming down to: this is the decision, and no data would make us change our minds.

In other words, i've heard a lot of concerns about things like perf.  But i haven't heard about any actual investigations indicating that hte perf concerns are real.  Indeed, the only actual bit of perf data provided seems to indicate there is no such perf concern here.  On the other hand, i've provided the arguments for my position, along with existence proofs of projects like TypeScript that demonstrate the feasibility of this pattern.  

There also doesn't seem to have been any effort whatsoever to determine the costs/implications of things like the updating all our API surface areas to pass this along.  It was brushed aside as " These will come later."  However, it seems exceptionally dangerous to try to make conclusions on the viability of this API with having done due diligence in these other areas.

Combined, it does not appear as if this proposal contains enough solid justification to warrant it (or, more specifically, to have it outweigh the alternatives).  If you want to take it, that is of course your call.  But i find it extremely surprising that this would be done without appropriate *measurements* of the varying potential approaches we could take here.

--

Another way of stating this:

Nullable went through more iterations than i can count in terms of the language-design *and* in terms of hte implementation strategy for the compiler.  We were willing to try out lots of things in order to learn and come up with the best possible approach.

However, for the API (which is arguably as important) we don't seem willing to do the same.  It seems liek the expectation is that we will just pick one, not measure perf or gauge the resultant API fallout, and that will be that.  

I honestly don't get this approach.  Roslyn has been willing many times in the past to experiment with complex APIs that will have a big impact on our ecosystem.  We did this with Syntax/SemanticModel.  We did this with our Analysis frameworks.  We did this with IOp.  We've done with this with IDE APIs.  So... why is it the case that we're just throwing away potential alternatives without a proper investigation as to the suitability and viability of all the proposals?

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 14 above)
Comment (CyrusNajmabadi) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: > TypeInfo is currently a light wrapper type around two ITypeSymbols. The IDE hangs onto these all the time, and TypeInfos don't have any references back to the context (read, position in a source tree) they came from. In order to calculate nullability information, you must have a location in a source tree, so these structs would suddenly no longer be lightweight wrappers.

IDE does not hang onto TypeInfos all the time.  Can you point me to examples that you're thinking of?    The XXXInfo guys are almost always ephemeral.  We do sometimes hold onto *Symbols* for shortish periods of time.  But even those we try to let go of since they can keep compilations alive.

Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: @jasonmalinowski or @DustinCampbell? I believe it was one of you who made that point about `TypeInfo` in the meetings. Regardless of that, we don't have interest in making `TypeInfo`s suddenly need to contain context to allow them to lazily calculate nullability.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 8 above)
Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: As an exercise here, I went though `SyntaxGenerator` and came up with a list of functions that would need to be added, how many usages of the function there are currently, and how many would be avoided if we added `Nullability` to `ITypeSymbol`. I did not include any unit test usages in this list, or usages within the syntax generator and subclasses.

```
Declarations
FieldDeclaration - 1 overload, none avoided. 2 uses.
MethodDeclaration - 1 overload, none avoided. 1 uses.
OperatorDeclaration - 1 overload, none avoided. 2 uses.
ParameterDeclaration - 1 overload, none avoided. 2 uses.
PropertyDeclaration - 1 overload, none avoided. No uses outside unit tests.
IndexerDeclaration - 1 overload, none avoided. No uses outside unit tests.
EventDeclarartion - 1 overload, none avoided. No uses outside unit tests.
CustomEventDeclaration - 1 overload, none avoided. No uses outside unit tests.
DelegateDeclaration - 1 overload, none avoided. No uses outside unit tests.
Declaration - 1 overload, 1 avoided. No uses outside unit tests.
AliasImportDirective - 1 overload, 1 avoided. 1 use.

Would want to add -
GetNullability
WithNullability

Statements
LocalDeclarationStatement - 5 new overloads. Only 1 would be eliminted by adding nullability to ITypeReference. 7 references, two of which would not need updating if ITypeSymbol had nullability
CatchClause - 2 overloads, none avoided, no uses

Expressions:
Worth noting here is that most of the usages are of the API that would be avoided.
DefaultExpression - 2 overloads, 1 avoided. 3 usages
IdentifierName - 2 overloads, none avoided. At least 90 usages outside unit tests
Identifer - 1 overload, none avoided. 11 usages
NameExpression - 1 overload, 1 avoided. 1 usages
TypeExpression - 3 overloads, 2 avoided. 43 usages
ArrayTypeExpression - 1 overload, none avoided. No uses outside unit tests
TupleTypeExpression - 1 overload, 1 avoided. No uses outside unit tests
TupleElementExpression - 2 overloads, 1 avoided. 2 uses
ArrayCreationExpression - 2 overloads, none avoided. No uses outside unit tests
ObjectCreationExpression - 4 overloads, 2 avoided. 9 uses
IsTypeExpression - 1 overload, 1 avoided. 4 uses
TryCastExpressio - 1 overload, 1 avoided. 3 uses.
CastExpression - 1 overload, 1 avoided. 11 uses.
ConvertExpression - 1 overload, 1 avoided. 1 use.
LambdaParameter - 1 overload, 1 avoided. No uses outside tests.
```

Comment (CyrusNajmabadi) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: Could you clarify the data?  I'm not sure i undertaand it. For example, can you break down what these mean: 

```
FieldDeclaration - 1 overload, none avoided. 2 uses.
...
TypeExpression - 3 overloads, 2 avoided. 43 usages
```

(i.e. actually go into the specific for these two cases).

Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: Sure.
`Field Declaration - needs 1 overload, would not have been avoided by adding Nullability to ITypeSymbol. 2 uses of one of the existing APIs not in test code or in the generator`
`Type Expression - needs 3 overloads, 2 would have been avoided by adding Nullability to ITypeSymbol. 43 uses of one of the existing APIs not in test code or in the generator`

Comment (CyrusNajmabadi) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: Sorry, i was asking for more detail than that.  Specifically:

> Field Declaration - needs 1 overload, would not have been avoided by adding Nullability to ITypeSymbol.

What overload would you intent to add?

Same with the others.  Can you explain the overlaods you intend?  I ask because this is whati see:

```c#
SyntaxNode FieldDeclaration(
            string name,
            SyntaxNode type,
            Accessibility accessibility = Accessibility.NotApplicable,
            DeclarationModifiers modifiers = default,
            SyntaxNode initializer = null);
SyntaxNode FieldDeclaration(IFieldSymbol field);
SyntaxNode FieldDeclaration(IFieldSymbol field, SyntaxNode initializer);
```

I don't see why any of those would need overloads added.  They all seem sufficient for handling the new NRT feature.  Thanks!

Comment (333fred) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: Yeah, you're right. Updated the declarations list to remove everything but `AliasImportDirective`, which will need the `AliasImportDeclaration(string aliasIdentifierName, INamespaceOrTypeSymbol symbol)` variant overloaded to take a `Nullability`, so you can say `using nullableType = Type?;`.

Comment (CyrusNajmabadi) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: > Yeah, you're right. Updated the declarations list to remove everything but AliasImportDirective

Why?  Since ITypeSymbol's would have the nullability info, that would come through `INamespaceOrTypeSymbol symbol`?

Comment (CyrusNajmabadi) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: > Updated the declarations list to remove everything

Why just the declarations list?  I'm not seeing why your other cases would need this either.  For example, from the 'expressions' i looked at, none seem like they would need overloads.  Can you clarify why you think they would?

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+This `SemanticModel` will share a cache with the nullable-aware `SemanticModel`, and all nullabilities accessed on symbols retrieved from this `SemanticModel` will be `NotAnalyzed`.
+This API will only be added if, after dogfooding, we determine that the performance of any scenarios is bad enough to warrant the introduction of the API.
+
+Generally speaking, when you want to determine the nullability of an `ITypeSymbol`, you need to check the containing context.
+The declaring symbol (e.g. `ILocalSymbol`, `IFieldSymbol`, etc.) will have the declared nullability of the type.
+`GetTypeInfo` will give you both the `ITypeSymbol` and the `Nullability` of an expression.
+Nested nullabilities for type parameters are contained on the containing `ITypeSymbol`.
+For example, if you wanted to know whether an array's contents are nullable, you will need to check the `IArrayTypeSymbol` for `ElementNullability`.
+
+Value types are always considered `NotNull`.
```
Comment (jcouv) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: nit: Except `Nullable<T>`.

File: docs/compilers/CSharp/Nullability Public API Design Notes.md [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+    Nullability Nullability { get; }
+}
+
+#endregion
+
+#region Flow State Nullability
+
+public interface IArrayTypeSymbol : ITypeSymbol
+{
+    Nullability ElementNullability { get; }
```
Comment (jcouv) [inline docs/compilers/CSharp/Nullability Public API Design Notes.md]: >ElementNullability [](start = 16, length = 18)

We probably need something for tuple element nullability as well

File: docs/compilers/CSharp/Nullability Public API Design Notes.md (hunk 6 above)
Comment (jmarolf) [reply docs/compilers/CSharp/Nullability Public API Design Notes.md]: I also don't see what scenario this is fixing.  The compiler needs to know about receiver/return nullability for its own tracking but this seems an odd way to expose this info.  Wouldn't the user use a flow analysis api in this case?

**Comment by jmarolf** [comment]
My main concern here is about how we want Equality to work.  My summary of this proposal is:

- All existing symbol equality checks remains unchanged
- New APIs are added for explicit nullabilty-symbol equality checks

I would need to see the ergonomics of these, but I would hope that we at lease add helpers for the second scenario.  Symbol equality has always been a hard thing from the analyzer author to get right.

**Comment by CyrusNajmabadi** [comment]
> I would need to see the ergonomics of these, but I would hope that we at lease add helpers for the second scenario. Symbol equality has always been a hard thing from the analyzer author to get right.

👍   IDE had to write an entire set of helpers for this themselves.

**Comment by 333fred** [comment]
I'm going to go ahead and merge this PR as it's the current POR: any follow up comments and design decisions can be added later.
